pkg/provisioning/provision.go

```diff
@@ -0,0 +1,240 @@
+// Copyright © 2022 Meroxa, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package provisioning
+
+import (
+	"context"
+	"io/ioutil"
+
+	"github.com/conduitio/conduit/pkg/connector"
+	"github.com/conduitio/conduit/pkg/foundation/log"
+	"github.com/conduitio/conduit/pkg/pipeline"
+	"github.com/conduitio/conduit/pkg/processor"
+)
+
+var update = false
+
+type Service struct {
+	ctx              *context.Context
+	logger           log.CtxLogger
+	pipelineService  *pipeline.Service
+	connectorService *connector.Service
+	processorService *processor.Service
+}
+
+func GetService(
+	ctx *context.Context,
+	logger log.CtxLogger,
+	plService *pipeline.Service,
+	connService *connector.Service,
+	procService *processor.Service,
+) *Service {
+	return &Service{
+		ctx:              ctx,
+		logger:           logger.WithComponent("provision"),
+		pipelineService:  plService,
+		connectorService: connService,
+		processorService: procService,
+	}
+}
+
+func (s *Service) ProvisionConfigFile(path string) error {
+	data, err := ioutil.ReadFile(path)
+	if err != nil {
+		return err
+	}
+
+	before, err := Parse(data)
+	if err != nil {
+		return err
+	}
+
+	got := EnrichPipelinesConfig(before)
+
+	for k, v := range got {
+		// skip non valid? multierror?
+		err = ValidatePipelinesConfig(v)
+		if err != nil {
+			return err
+		}
+		err = s.provisionPipeline(k, v)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (s *Service) provisionPipeline(id string, config PipelineConfig) error {
```

Comment:
It would be great if this operation would happen inside a transaction and rollback any changes if it didn't succeed, otherwise we risk ending up with a partially provisioned pipeline. For the transaction `Service` will need access to the db, for rollbacks there is the utility `rollback.R`. See the [orchestrator](https://github.com/ConduitIO/conduit/blob/30a574c86c10d9280517a0ea086339dd8bbd5662/pkg/orchestrator/processors.go#L35-L42) how we use it there.